services/ui/src/components/Projects/index.js

```diff
@@ -52,7 +52,7 @@ const Projects = ({ projects = [] }) => {
           className="searchInput"
           type="text"
           value={searchInput}
-          onChange={e => setSearchInput(e.target.value)}
+          onChange={e => setSearchInput(e.target.value.replace(/[^0-9a-zA-Z.-]/g,''))}
```

Comment:
Real-time frontend frameworks like React allow us to do things that weren't really possible before and this is a clever and easy technical solution to the problem, which is why I've tried it on past projects. Unfortunately it results in a bad user experience for form input.

This implementation creates a militant type of form validation where invalid user input is ignored. But to the user whet they see is that they'll type an `_` but nothing on the page changes. The way 99% of websites do form validation is by telling the user what they did wrong, and many of the HTML5 client-side validations still allow invalid input. To the user, it will seem like the site is broken, not that we are validating their input (during UAT we had one user ask for a spare keyboard because they thought a key was broken).

IMHO the bare minimum to close the original issue would be to add a new `onPaste` handler that trims whitespace only on paste.

The current form doesn't use any form validation at all, we just depend on the "no results" message. If we wanted to change that we could implement full/proper form validation by showing error messages and error styling with the underlying principle that user input must always be kept (including if the user types a `space`).

Reply:
Really solid points, makes sense! I've briefly discussed this with Blaize and instead of mutating the input value, we're now trimming the whitespace in the filtering step. This won't restrict user input like the previous iteration, but will ignore invalid whitespace.

Reply:
Thanks, LGTM!